src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                        : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, selectedItemIndex, filters,
+                updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleDeletionTrigger(
+            List<ExtendedFilterResult> filterResults,
+            string filterText,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight> highlightedList)
+        {
+            ExtendedFilterResult? bestFilterResult = null;
+            int matchCount = 0;
+            foreach (var currentFilterResult in filterResults.Where(r => r.FilterResult.MatchedFilterText))
+            {
+                if (bestFilterResult == null ||
+                    Session.IsBetterDeletionMatch(currentFilterResult.FilterResult, bestFilterResult.Value.FilterResult))
+                {
+                    // We had no best result yet, so this is now our best result.
+                    bestFilterResult = currentFilterResult;
+                    matchCount++;
+                }
+            }
+
+            int index;
+            bool hardSelect;
+
+            // If we had a matching item, then pick the best of the matching items and
+            // choose that one to be hard selected.  If we had no actual matching items
+            // (which can happen if the user deletes down to a single character and we
+            // include everything), then we just soft select the first item.
+            if (bestFilterResult != null)
+            {
+                // Only hard select this result if it's a prefix match
+                // We need to do this so that
+                // * deleting and retyping a dot in a member access does not change the
+                //   text that originally appeared before the dot
+                // * deleting through a word from the end keeps that word selected
+                // This also preserves the behavior the VB had through Dev12.
+                hardSelect = bestFilterResult.Value.VSCompletionItem.FilterText.StartsWith(filterText, StringComparison.CurrentCultureIgnoreCase);
+                index = filterResults.IndexOf(bestFilterResult.Value);
+            }
+            else
+            {
+                index = 0;
+                hardSelect = false;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, index, filters,
+                hardSelect ? AsyncCompletionData.UpdateSelectionHint.Selected : AsyncCompletionData.UpdateSelectionHint.SoftSelected,
+                centerSelection: true, uniqueItem: null);
```

Comment:
we used to have:

``.WithIsUnique(matchCount == 1);``

is it correct to pass 'null' for uniqueItem here?  What if it is unique?